store.go

```diff
@@ -40,7 +45,91 @@ func (t *TableBlock) Persist() error {
 	return nil
 }
 
-func (t *Table) IterateBucketBlocks(ctx context.Context, logger log.Logger, filter TrueNegativeFilter, iterator func(rg dynparquet.DynamicRowGroup) bool, lastBlockTimestamp uint64) error {
+// BlockFilterFunc takes a block ULID and returns true if this block can be ignored during this query.
+type BlockFilterFunc func(ulid.ULID, *parquet.File) bool
+
+// A BlockFilter is a logical construction of multiple block filters.
+type BlockFilter struct {
+	BlockFilterFunc
+	input *BlockFilter
+}
+
+// Filter will recursively call input filters until one returns true or the final filter is reached.
+func (b *BlockFilter) Filter(block ulid.ULID, file *parquet.File) bool {
+	if b.BlockFilterFunc == nil {
+		return false
+	}
+
+	if b.BlockFilterFunc(block, file) {
+		return true
+	}
+
+	return b.input.Filter(block, file)
+}
+
+// LastBlockTimestamp filters blocks that are still being written or still held in memory.
+func (b *BlockFilter) LastBlockTimestamp(lastBlockTimestamp uint64) *BlockFilter {
+	return &BlockFilter{
+		BlockFilterFunc: func(block ulid.ULID, file *parquet.File) bool {
+			return lastBlockTimestamp != 0 && block.Time() >= lastBlockTimestamp
+		},
+		input: b,
+	}
+}
+
+func (b *BlockFilter) MinMaxFilter(filter logicalplan.Expr) *BlockFilter {
+	return &BlockFilter{
+		BlockFilterFunc: func(block ulid.ULID, file *parquet.File) bool {
+			mins := map[string]*parquet.Value{}
+			maxes := map[string]*parquet.Value{}
+			for i, cidx := range file.ColumnIndexes() {
+				var c parquet.ColumnIndex
+				switch *file.Metadata().Schema[i+1].Type { // TODO THOR this is probably not the correct way to do this
```

Comment:
Feels like there has to be a better way than what I'm doing here...

Reply:
Do you mean the way you're retrieving the metadata or this switch statement for the types themselves?